libazureinit/src/config.rs

```diff
@@ -0,0 +1,820 @@
+///
+/// Configures telemetry behavior, including diagnostic settings.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Telemetry {
+    /// Flag to enable or disable KVP diagnostics. Defaults to `true`.
+    pub kvp_diagnostics: bool,
+}
+
+impl Default for Telemetry {
+    fn default() -> Self {
+        Self {
+            kvp_diagnostics: true,
+        }
+    }
+}
+
+/// General configuration struct for azure-init.
+///
+/// Aggregates all configuration settings for managing SSH, provisioning, IMDS, media,
+/// and telemetry, supporting loading from file or directory and merging configurations.
+#[derive(Default, Serialize, Deserialize, Debug, Clone)]
+#[serde(default)]
+pub struct Config {
+    pub ssh: Ssh,
+    pub hostname_provisioners: HostnameProvisioners,
+    pub user_provisioners: UserProvisioners,
+    pub password_provisioners: PasswordProvisioners,
+    pub imds: Imds,
+    pub provisioning_media: ProvisioningMedia,
+    pub azure_proxy_agent: AzureProxyAgent,
+    pub wireserver: Wireserver,
+    pub telemetry: Telemetry,
+}
+
+/// Implements `Display` for `Config`, formatting it as a readable TOML string.
+///
+/// Uses `toml::to_string_pretty` to serialize the configuration. If serialization fails,
+/// a fallback message is displayed..
+impl fmt::Display for Config {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            toml::to_string_pretty(self)
+                .unwrap_or_else(|_| "Unable to serialize config.".to_string())
+        )
+    }
+}
+
+/// Loads the configuration for `azure-init`.
+///
+/// This method uses the `Figment` library to load configuration from the following sources,
+/// in order of priority:
+///
+/// 1. **Defaults**: Base configuration from `Config::default()`.
+/// 2. **Main File**: `azure-init.toml`, if present.
+/// 3. **Directory Files**: `.toml` files in `azure-init.d`, sorted lexicographically.
+/// 4. **CLI Overrides**: A file or directory specified via the CLI.
+///
+/// Later sources override earlier ones in case of conflicts.
+impl Config {
+    pub fn load(path: Option<PathBuf>) -> Result<Config, Error> {
+        let mut figment = Figment::new()
+            .merge(Serialized::defaults(Config::default()))
+            .merge(Toml::file("azure-init.toml"));
+
+        if let Some(path) = path {
+            if path.is_dir() {
+                let mut entries: Vec<_> = fs::read_dir(&path)
+                    .map_err(|e| {
+                        tracing::error!("Failed to read directory: {:?}", e);
+                        Error::Io(e)
+                    })?
+                    .filter_map(Result::ok)
+                    .map(|entry| entry.path())
+                    .filter(|path| {
+                        path.extension().map_or(false, |ext| ext == "toml")
+                    })
+                    .collect();
+
+                entries.sort();
+
+                for path in entries {
```

Comment:
perhaps `for path_entry in entries` so that we don't override the same variable usage "path" (which is used in the let statement)